docker-app/qfieldcloud/notifs/models.py

```diff
@@ -0,0 +1,260 @@
+"""
+This module logs activity streams using the `django-activity-stream` package.
+"""
+from django.db.models.signals import post_save, pre_delete
+from django.dispatch import receiver
+from django_currentuser.middleware import get_current_authenticated_user
+from notifications.signals import notify
+
+from ..core.models import (
+    Organization,
+    OrganizationMember,
+    Project,
+    ProjectCollaborator,
+    Team,
+    TeamMember,
+    User,
+)
+
+
+def _send_notif(verb, action_object, recipient, target=None):
+    """
+    Sends a notification through django-notifications, but sets it as already read
+    if the actor is the same as the recipient
+    """
+    authenticated_user = get_current_authenticated_user()
+
+    # django-notifications-hq doesn't support Anonymous users yet (e.g. management command)
+    if authenticated_user is None:
+        return
+
+    # deduplicate
+    recipient = list(set(recipient))
+
+    if authenticated_user in recipient:
+        # If the actor is already the recipient, we consider the notification as seen
+
+        # TODO : marking notifications as read is currently not supported, so we disable
+        # self-notifications for now
+        # see https://github.com/django-notifications/django-notifications/issues/317
+        # notify.send(
+        #     authenticated_user,
+        #     verb=verb,
+        #     action_object=action_object,
+        #     target=target,
+        #     recipient=[authenticated_user],
+        #     unread=False, # not supported !
+        # )
+        recipient.remove(authenticated_user)
+
+    notify.send(
+        authenticated_user,
+        verb=verb,
+        action_object=action_object,
+        target=target,
+        recipient=list(recipient),
+    )
+
+
+def _subusers_for_user(user: User):
+    """Returns a list of users (of TYPE_USER) concerned by updates to an user (any type)"""
+
+    if user.user_type == User.TYPE_USER:
+        return [user]
+    elif user.user_type == User.TYPE_TEAM:
+        team_members = []
+        for m in user.team.members.all():
+            team_members.extend(_subusers_for_user(m.member))
+        organization_admins = _subusers_for_user(
+            user.team.team_organization.organization_owner
+        )
+        return team_members + organization_admins
+    elif user.user_type == User.TYPE_ORGANIZATION:
+        org_members = []
+        for m in user.organization.members.all():
+            org_members.extend(_subusers_for_user(m.member))
+        org_owners = _subusers_for_user(user.organization.organization_owner)
+        return org_members + org_owners
+
+
+def _users_for_project(project: Project):
+    """Returns a list of users concerned by updates to a project"""
+
+    proj_members = []
+    for m in project.collaborators.all():
+        proj_members.extend(_subusers_for_user(m.collaborator))
+    proj_owners = _subusers_for_user(project.owner)
+    return proj_members + proj_owners
+
+
+# Register signals
```

Comment:
Can we move all these signals into `signals.py`? So it's cleaner when we add more signals to listen for changes or modify the model.